### PR TITLE
Prevent archiving / restoring questions that aren't in a valid state.

### PR DESCRIPTION
**Pull request by clouser2**
### Description

Presently, the `/archive` and `/restore` handlers don't do any validation beyond the question ID existing. This means that a question that is still referenced by a program can be archived.

This could definitely occur for a stale tab scenario where an admin has the archive option, then references the question from a separate program (rendering it unarchivable).

Additionally, we should always rely on validation before we mutate state rather than relying on links being generated at the appropriate time.

Currently, this returns a 400 error. As part of #2788, I may adjust the controller to redirect to the main question page with an error message.

## Release notes:

Fix issue where UI could allow accidental archival of questions that were still referenced.

### Checklist

- [X] Created tests which fail without the change (if possible)

### Issue(s)

#2788

### CODE REVIEW

File: server/app/services/question/QuestionServiceImpl.java [hunk 1]
```diff
@@ -112,6 +116,12 @@ public void restoreQuestion(Long id) throws InvalidUpdateException {
     if (question.isEmpty()) {
       throw new InvalidUpdateException("Did not find question.");
     }
+    ActiveAndDraftQuestions activeQuestions =
+        readOnlyQuestionService().getActiveAndDraftQuestions();
+    if (activeQuestions.getDeletionStatus(question.get().getQuestionDefinition().getName())
+        != DeletionStatus.PENDING_DELETION) {
```
Comment (bion) [inline server/app/services/question/QuestionServiceImpl.java:122]: suggestion (blocking): use .equals for enums

File: server/app/services/question/QuestionServiceImpl.java [hunk 2]
```diff
@@ -126,6 +136,12 @@ public void archiveQuestion(Long id) throws InvalidUpdateException {
     if (question.isEmpty()) {
       throw new InvalidUpdateException("Did not find question.");
     }
+    ActiveAndDraftQuestions activeQuestions =
+        readOnlyQuestionService().getActiveAndDraftQuestions();
+    if (activeQuestions.getDeletionStatus(question.get().getQuestionDefinition().getName())
+        != DeletionStatus.DELETABLE) {
```
Comment (bion) [inline server/app/services/question/QuestionServiceImpl.java:142]: suggestion (blocking): use .equals for enums

**Review by bion — CHANGES_REQUESTED** [review]
Does the UI allow a way of doing this? If so we should update the UI as well

**Comment by clouser2** [comment]
> Does the UI allow a way of doing this? If so we should update the UI as well

Not presently assuming everything occurs in a single tab. Multi-tab scenarios would allow this though. In a future PR, I'll likely redirect to the question page with an error message in the flash.

File: server/app/services/question/QuestionServiceImpl.java (hunk 1 above)
Comment (clouser2) [reply server/app/services/question/QuestionServiceImpl.java]: Done.

File: server/app/services/question/QuestionServiceImpl.java (hunk 2 above)
Comment (clouser2) [reply server/app/services/question/QuestionServiceImpl.java]: Done.

**Review by bion — APPROVED** [review]

**Pull request merged**